docs/get-details/dapps/smart-contracts/abi/index.md

```diff
@@ -0,0 +1,77 @@
+
+
+# DataTypes
+
+Encoding for the datatypes is specified [here](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md#encoding)
+
+The datatypes specified are: 
+
+| Type | Description | 
+| ---- | ----------- |
+|uintN| An N-bit unsigned integer, where 8 <= N <= 512 and N % 8 = 0|
+|byte| An alias for uint8|
+|bool| A boolean value that is restricted to either 0 or 1. When encoded, up to 8 consecutive bool values will be packed into a single byte|
+|ufixedNxM| An N-bit unsigned fixed-point decimal number with precision M, where 8 <= N <= 512, N % 8 = 0, and 0 < M <= 160, which denotes a value v as v / (10^M)|
+|type[N]| A fixed-length array of length N, where N >= 0. type can be any other type|
+|address| Used to represent a 32-byte Algorand address. This is equivalent to byte[32]|
+|type[]| A variable-length array. type can be any other type|
+|string| A variable-length byte array (byte[]) assumed to contain UTF-8 encoded content|
+|(T1,T2,...,TN)| A tuple of the types T1, T2, …, TN, N >= 0|
+|reference type | account, asset, application only for arguments, in which case they are an alias for uint8. See section "Reference Types" below|
+
+The encoding and decoding of these types should be handled by the SDKs for calling methods and reading out return values.  
+
+Because stack types in the AVM are limited to uint64 and bytes, a smart contract may rely on ABI types defined in PyTeal to encode or decode the data passed in the application args. 
+
+## Reference Types
+
+Reference types may be specified in the method signature referring to some transaction parameters that must be passed.  The value encoded is a uint8 reference to the index of element in the relevant array (i.e. for account, the index in the foreign accounts array)
+
+
+# Methods
+
+Methods may be exposed by the smart contract and called by submitting an ApplicationCall transaction to the existing application id. 
+
+A *method signature* is defined as a name, argument types, and return type. The stringified version is then hashed and the first 4 bytes are taken as a *method selector*
+
+For example:
+
+A *method signature* for an `add` method that takes 2 uint64s and returns 1 uint128
+```
+Method signature: add(uint64,uint64)uint128
+```
+
+The string version of the *method signature* is hashed and the first 4 bytes are its *method selector*
+```
+SHA-512/256 hash (in hex): 8aa3b61f0f1965c3a1cbfa91d46b24e54c67270184ff89dc114e877b1753254a
+Method selector (in hex): 8aa3b61f
+```
+
+Once the method selector is known, it is used in the smart contract logic to route to the appropriate logic that implements the `add` method. 
+
+The `method` pseudo-opcode can be used in a contract to do the above work and produce a *method selector* given the *method signature* string.
+
+```
+method "add(uint64,uint64)uint64"
```

Comment:
I just realised this method has a return value of uint64 rather than uint128. Doesn't really matter as it's not an example, but could be consistent with the above.